### DISCUSSION TITLE
Allow styling descendant of group that matches peer pseudo class variant

### DISCUSSION THREAD

**Discussion by wolthers** [tailwindlabs/tailwindcss#8777] [Ideas]
### Usecase:
When writing a custom checkbox or radio component it is often desirable to be able to style children of a group depending on the state of a sibling of the group.

### Proposed implementation:
```
<div>
  <input class="peer" type="radio" />
  <label class="group">
    <span class="group-peer-checked:hidden">☐</span>
    <span class="hidden group-peer-checked:block">☑</span>
    Some label
  </label>
</div>
```

Currently solvable with a plugin like this, but usecase is common enough that I think it makes sense to integrate.

```javascript
// tailwind.config.js
module.exports = {
  ....
  plugins: [
    plugin(function groupPeer({ addVariant }) {
      let pseudoVariants = [
        // ... Any other pseudo variants you want to support. 
        // See https://github.com/tailwindlabs/tailwindcss/blob/6729524185b48c9e25af62fc2372911d66e7d1f0/src/corePlugins.js#L78
        "checked",
      ].map((variant) =>
        Array.isArray(variant) ? variant : [variant, `&:${variant}`],
      );

      for (let [variantName, state] of pseudoVariants) {
        addVariant(`group-peer-${variantName}`, (ctx) => {
          let result = typeof state === "function" ? state(ctx) : state;
          return result.replace(/&(\S+)/, ":merge(.peer)$1 ~ .group &");
        });
      }
    }),
  ],
  ....
};
```

If I understand the source well enough, it should be relatively straight forward to add in the pseudoClass variants block here: https://github.com/tailwindlabs/tailwindcss/blob/6729524185b48c9e25af62fc2372911d66e7d1f0/src/corePlugins.js#L77

**Comment by pofulu** [comment]
Thank you! It's really useful in many case, hope it become officially built-in api.

**Comment by maxdzin** [comment]
Thanks, @wolthers 
That's helped me too!
And I also hope that we can see it built-in officially.

**Comment by ZackPlauche** [comment]
I hope they do add this, seriously.

Also, to get this to work with [Nuxt 3](), I had to write it without the `plugin` function:
```js
// tailwind.config.js
module.exports = {
  ....
  plugins: [
    function groupPeer({ addVariant }) {
      let pseudoVariants = [
        // ... Any other pseudo variants you want to support. 
        // See https://github.com/tailwindlabs/tailwindcss/blob/6729524185b48c9e25af62fc2372911d66e7d1f0/src/corePlugins.js#L78
        "checked",
      ].map((variant) =>
        Array.isArray(variant) ? variant : [variant, `&:${variant}`],
      );

      for (let [variantName, state] of pseudoVariants) {
        addVariant(`group-peer-${variantName}`, (ctx) => {
          let result = typeof state === "function" ? state(ctx) : state;
          return result.replace(/&(\S+)/, ":merge(.peer)$1 ~ .group &");
        });
      }
    },
  ],
  ...
};
```

**Comment by georgfaust** [comment]
👍🏻 

we need sth to implmenent this standard pattern

```
<style>
  .content { display: none; }
  #aaa:checked ~ #parent .aaa,
  #bbb:checked ~ #parent .bbb { background-color: green; }
  #aaa:checked ~ #children .aaa { display: block; }
  #bbb:checked ~ #children .bbb { display: block; }
</style>
<body>
  <input type="radio" id="aaa" name="content-filter" checked />
  <input type="radio" id="bbb" name="content-filter" />

  <div id="parent">
    <label for="aaa" class="aaa" id="aaa-label">AAA</label>
    <label for="bbb" class="bbb" id="bbb-label">BBB</label>
  </div>

  <div id="children">
    <div class="content aaa">aaa</div>
    <div class="content bbb">bbb</div>
  </div>
</body>
```

**Comment by MurmeltierS** [comment]
If I'm not mistaken, you can achieve this by using `peer-checked:group-[]:`.

example:
```html
<input type="checkbox" class="peer" />
<div class="text-blue-400 peer-checked:text-red-500 group">hello
  <div></div>
  <div class="peer-checked:group-[]:hidden">Icon</div>
</div>
```
https://play.tailwindcss.com/NT2vsY9Al5

**Update:**
In Tailwind 4.0+ you're able to use `group-peer-`. The above workaround won't work anymore.

```html
<input type="checkbox" class="peer" />
<div class="text-blue-400 peer-checked:text-red-500 group">hello
  <div></div>
  <div class="group-peer-checked:hidden">Icon</div>
</div>
```
https://play.tailwindcss.com/dJE3pNQH8Z

**Reply by tleunen** [reply]
How do you make this work when the group is named?

**Reply by MurmeltierS** [reply]
@tleunen just like this

```html
<input type="checkbox" class="peer" />
<div class="text-blue-400 peer-checked:text-red-500 group/myname">hello
  <div></div>
  <div class="peer-checked:group-[]/myname:hidden">Icon</div>
</div>
```

https://play.tailwindcss.com/STpVNa2s7y

**Reply by Jorundur** [reply]
@MurmeltierS What do the empty square brackets signify in `group-[]`? Why can't we just do `group`?

**Reply by wongjn** [reply]
~~`group` signifies a parent element. `group-[]` means we are *depending* on a parent `group`.~~

`group` needs some sort state/variant/modifier attached to it. The `[]` gives satisfies this requirement.

**Reply by adamwathan** [reply]
Even I am surprised that this works 😂

**Reply by vamirio256** [reply]
This works! Thanks very much.

**Reply by ZackPlauche** [reply]
> Even I am surprised that this works 😂

Could you guys make it into something more straightforward lol

**Reply by ChazUK** [reply]
Amazing!

**Reply by i1ko** [reply]
Tnx!!!

But why my below variant doesn't work?

```
<input type="checkbox" class="peer" />
<div class="text-blue-400 peer-checked:text-red-500 peer-checked:is-checked group">hello
  <div></div>
  <div class="group-[.is-checked]:hidden">Icon</div>
</div>
```

**Reply by MurmeltierS** [reply]
@i1ko Your tailwind class `group-[.is-checked]:hidden` translates to: 
_If the **group** element has the class **.is-checked**, then turn the Icon **hidden**._

So your icon would be (always) hidden if you would add the `is-checked` class to your group div.
```diff
-  <div class="text-blue-400 peer-checked:text-red-500 peer-checked:is-checked group">
+  <div class="text-blue-400 peer-checked:text-red-500 peer-checked:is-checked group is-checked">
```

This wouldn't make much sense though, as the icon would always be hidden - regardless of the checkbox.

**Reply by yersskit** [reply]
you deserve some beers, it works!

**Reply by excidence** [reply]
In Tailwind 4.0 beta you'll simply need to replace `peer-checked:group-[]:hidden` with `group-peer-checked:hidden` for this to work again 😁

**Reply by TCheeMing** [reply]
How would this work with a named peer?
